README, CI: adjust for the WASI toolchain rename

piodul opened the pull request:
The wasm32-wasi toolchain is being renamed to wasm32-wasip1 and the old name will no longer will be accepted starting from Rust 1.84. Prepare for the change by adjusting the existing references to it:

- README now uses the new name, and a note was introduced which mentions that the toolchain was renamed at some point and older versions of Rust might still use the old name.
- The rust.yml workflow is updated to use wasm32-wasip1, except for the MSRV worfklow which deliberately uses an old version of Rust; that version does not support the new toolchain name.

In addition, dependencies for the MSRV run are pinned so that the MSRV CI job passes.

Fixes: scylladb/scylla-rust-udf#16

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
-->

- [x] I have split my patch into logically separate commits.
- [ x All commit messages clearly explain what they change and why.
- [ ] ~I added relevant tests for new features and bug fixes.~
- [ ] All commits compile, pass static checks and pass tests.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

wmitros commented on .github/workflows/rust.yml:
```diff
@@ -18,27 +18,28 @@ jobs:
         curl https://wasmtime.dev/install.sh -sSf | bash
     - name: Test
-      run: CARGO_TARGET_WASM32_WASI_RUNNER="/home/runner/.wasmtime/bin/wasmtime --allow-unknown-exports" cargo test --target=wasm32-wasi --all-targets
+      run: CARGO_TARGET_WASM32_WASI_RUNNER="/home/runner/.wasmtime/bin/wasmtime --allow-unknown-exports" cargo test --target=wasm32-wasip1 --all-targets
```
```suggestion
      run: CARGO_TARGET_WASM32_WASIP1_RUNNER="/home/runner/.wasmtime/bin/wasmtime --allow-unknown-exports" cargo test --target=wasm32-wasip1 --all-targets
```

piodul replied:
I guessed that it could be a fix for the failure, I already tried it locally but I get the same result when I run it on my machine... Does it work on yours?

wmitros replied:
Looks like it doesn't work on mine either - the `--allow-unknown-exports` was changed and it looks like right now we need to use `-W unknown-exports-allow=y` instead (it works on my machine with that

piodul replied:
Thanks, it now seems to work on my machine, too.

(I also forgot to substitute the `runner` for my username, I changed the invocation to use `$HOME` to prevent such mistakes in the future)

piodul commented:
v2:
- Fixed one path previously unfixed in v1 - which looked incorrect before the change, fixed that as well
- Changed `/home/runner` to `$HOME`
- Changed the `--allow-unknown-exports` flag to the current version of the flag, accepted by the newest runtime

piodul commented:
The `test` job passes, but it looks like we didn't pin our dependencies and the MSRV job fails. I will fix that later (this PR will have to wait for this).

piodul commented:
v3: included a commit which pins dependencies in the MSRV CI job
